titus-server-gateway/src/main/java/com/netflix/titus/gateway/service/v3/internal/GatewayJobServiceGateway.java

```diff
@@ -294,17 +293,21 @@ public GatewayJobServiceGateway(GrpcRequestConfiguration tunablesConfiguration,
     @Override
     public Observable<JobChangeNotification> observeJob(String jobId, CallMetadata callMetadata) {
         if (localCacheQueryProcessor.canUseCache(Collections.emptyMap(), "observeJob", callMetadata)) {
-            return localCacheQueryProcessor.syncCache("observeJob", JobChangeNotification.class).concatWith(localCacheQueryProcessor.observeJob(jobId));
+            return localCacheQueryProcessor.syncCache("observeJob", JobChangeNotification.class)
+                    .concatWith(localCacheQueryProcessor.observeJob(jobId))
+                    .map(taskRelocationDataInjector::injectIntoTaskUpdateEvent);
         }
-        return super.observeJob(jobId, callMetadata);
+        return super.observeJob(jobId, callMetadata).map(taskRelocationDataInjector::injectIntoTaskUpdateEvent);
```

Comment:
I've not seen this code before, but before, this localCacheQueryProcessor thing was doing the heavy lifting, regardless of the number of observeJobs streamers.

But now with the `.map()`, will we be calling (or really, looking at the shared informer) for each job and for each streamer? If so, any worries about scaling here?

Reply:
`localCacheQueryProcessor` provides the raw data. The data injector decorates them with additional information.
The shared informer calls should be inexpensive, just like setting the data in the protobuf messages.